[WOOMOB-946] Remove deprecated testing approach

malinajirka opened the pull request:
Closes https://github.com/woocommerce/woocommerce-android/issues/6899
Close WOOMOB-946

<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

Goal of this PR is to remove the long deprecated approach to testing coroutines. I originally planned to update the tests one by one, but ultimately decided to completely rewrite them and add a few more tests with the help of AI.

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

Green CI is enough

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
N/A

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

malinajirka committed:
Remove deprecated approach for handling coroutines in tests

malinajirka committed:
Remove deprecated usage in ReviewModerationHandlerTests

ParaskP7 commented:
Nice @malinajirka and thanks ❤️ this actually then closes #6899 and we can add that to the description to have that auto-closed when this gets merged, I am on-it with the review... 👀

ParaskP7 reviewed (approved):
👋 @malinajirka , the tests change LGTM, and thus I am unblocking this with a ✅ no matter, many thanks! 🥇

Also, I tried making the test suite a bit more robust, by initializing `handler` once and using `TestScope(coroutinesTestRule.testDispatcher)` instead of a `backgroundScope` scope on it. See patch below, let me know what you think! 🙏 

<details>
    <summary>patch</summary>

```diff
Index: WooCommerce/src/test/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationHandlerTests.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/WooCommerce/src/test/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationHandlerTests.kt b/WooCommerce/src/test/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationHandlerTests.kt
--- a/WooCommerce/src/test/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationHandlerTests.kt	(revision 91fc5d42ed6d1f357b98a0ae358ef7b3170a9941)
+++ b/WooCommerce/src/test/kotlin/com/woocommerce/android/ui/reviews/ReviewModerationHandlerTests.kt	(date 1753964076107)
@@ -2,14 +2,11 @@
 
 import com.woocommerce.android.model.ActionStatus
 import com.woocommerce.android.tools.SelectedSite
-import com.woocommerce.android.ui.reviews.ProductReviewStatus.APPROVED
-import com.woocommerce.android.ui.reviews.ProductReviewStatus.HOLD
-import com.woocommerce.android.ui.reviews.ProductReviewStatus.SPAM
-import com.woocommerce.android.ui.reviews.ProductReviewStatus.TRASH
 import com.woocommerce.android.viewmodel.BaseUnitTest
 import kotlinx.coroutines.ExperimentalCoroutinesApi
 import kotlinx.coroutines.flow.first
 import kotlinx.coroutines.launch
+import kotlinx.coroutines.test.TestScope
 import kotlinx.coroutines.test.advanceTimeBy
 import kotlinx.coroutines.test.runCurrent
 import kotlinx.coroutines.test.runTest
@@ -38,7 +35,11 @@
 
     private val review = ProductReviewTestUtils.generateProductReview(0L, 0L)
 
-    private lateinit var handler: ReviewModerationHandler
+    private val handler = ReviewModerationHandler(
+        selectedSite = selectedSite,
+        productStore = productStore,
+        appCoroutineScope = TestScope(coroutinesTestRule.testDispatcher),
+    )
 
     @Before
     fun setup() = runTest {
@@ -46,7 +47,7 @@
             WooResult(
                 WCProductReviewModel(
                     remoteProductReviewId = RemoteId(review.remoteId),
-                    status = HOLD.toString(),
+                    status = ProductReviewStatus.HOLD.toString(),
                 )
             )
         )
@@ -56,11 +57,6 @@
     fun `when moderating a review, then start with a pending state`() = runTest {
         // GIVEN
         val statusUpdates = mutableListOf<List<ReviewModerationStatus>>()
-        handler = ReviewModerationHandler(
-            selectedSite = selectedSite,
-            productStore = productStore,
-            appCoroutineScope = backgroundScope,
-        )
 
         val collectJob = launch {
             handler.pendingModerationStatus.collect { statuses ->
@@ -70,7 +66,7 @@
         runCurrent()
 
         // WHEN
-        handler.postModerationRequest(review, HOLD)
+        handler.postModerationRequest(review, ProductReviewStatus.HOLD)
         runCurrent()
 
         // THEN
@@ -86,11 +82,6 @@
     fun `given moderating a review, when the undo delay is passed, then change status to submitted`() = runTest {
         // GIVEN
         val statusUpdates = mutableListOf<List<ReviewModerationStatus>>()
-        handler = ReviewModerationHandler(
-            selectedSite = selectedSite,
-            productStore = productStore,
-            appCoroutineScope = backgroundScope,
-        )
 
         val collectJob = launch {
             handler.pendingModerationStatus.collect { statuses ->
@@ -99,7 +90,7 @@
         }
 
         runCurrent()
-        handler.postModerationRequest(review, HOLD)
+        handler.postModerationRequest(review, ProductReviewStatus.HOLD)
         runCurrent()
 
         val pendingStatuses = statusUpdates.flatten().filter { it.actionStatus == ActionStatus.PENDING }
@@ -124,11 +115,6 @@
     fun `given moderating a review, when the undo delay is passed, then submit status to the API`() = runTest {
         // GIVEN
         val statusUpdates = mutableListOf<List<ReviewModerationStatus>>()
-        handler = ReviewModerationHandler(
-            selectedSite = selectedSite,
-            productStore = productStore,
-            appCoroutineScope = backgroundScope,
-        )
 
         val collectJob = launch {
             handler.pendingModerationStatus.collect { statuses ->
@@ -137,7 +123,7 @@
         }
 
         runCurrent()
-        handler.postModerationRequest(review, HOLD)
+        handler.postModerationRequest(review, ProductReviewStatus.HOLD)
         runCurrent()
 
         // WHEN
@@ -154,14 +140,8 @@
     @Test
     fun `when review status update succeeds, then status removed after a given time`() = runTest {
         // GIVEN
-        handler = ReviewModerationHandler(
-            selectedSite = selectedSite,
-            productStore = productStore,
-            appCoroutineScope = backgroundScope,
-        )
-
         runCurrent()
-        handler.postModerationRequest(review, HOLD)
+        handler.postModerationRequest(review, ProductReviewStatus.HOLD)
         runCurrent()
 
         advanceTimeBy(ReviewModerationHandler.UNDO_DELAY + 100)
@@ -184,11 +164,6 @@
         )
 
         val statusUpdates = mutableListOf<List<ReviewModerationStatus>>()
-        handler = ReviewModerationHandler(
-            selectedSite = selectedSite,
-            productStore = productStore,
-            appCoroutineScope = backgroundScope,
-        )
 
         val collectJob = launch {
             handler.pendingModerationStatus.collect { statuses ->
@@ -197,7 +172,7 @@
         }
 
         runCurrent()
-        handler.postModerationRequest(review, HOLD)
+        handler.postModerationRequest(review, ProductReviewStatus.HOLD)
         runCurrent()
 
         // WHEN
@@ -218,14 +193,8 @@
             WooResult(WooError(GENERIC_ERROR, UNKNOWN, ""))
         )
 
-        handler = ReviewModerationHandler(
-            selectedSite = selectedSite,
-            productStore = productStore,
-            appCoroutineScope = backgroundScope,
-        )
-
         runCurrent()
-        handler.postModerationRequest(review, HOLD)
+        handler.postModerationRequest(review, ProductReviewStatus.HOLD)
         runCurrent()
 
         advanceTimeBy(ReviewModerationHandler.UNDO_DELAY + 100)
@@ -243,14 +212,8 @@
     @Test
     fun `when moderation is canceled, then submit success with original status`() = runTest {
         // GIVEN
-        handler = ReviewModerationHandler(
-            selectedSite = selectedSite,
-            productStore = productStore,
-            appCoroutineScope = backgroundScope,
-        )
-
         runCurrent()
-        handler.postModerationRequest(review, HOLD)
+        handler.postModerationRequest(review, ProductReviewStatus.HOLD)
         runCurrent()
 
         val pendingStatus = handler.pendingModerationStatus.first()
@@ -271,11 +234,6 @@
     fun `when queuing a second moderation request, then skip delay for the previous one`() = runTest {
         // GIVEN
         val statusUpdates = mutableListOf<List<ReviewModerationStatus>>()
-        handler = ReviewModerationHandler(
-            selectedSite = selectedSite,
-            productStore = productStore,
-            appCoroutineScope = backgroundScope,
-        )
 
         val review1 = review.copy(remoteId = 1L)
         val review2 = review.copy(remoteId = 2L)
@@ -287,12 +245,12 @@
         }
 
         runCurrent()
-        handler.postModerationRequest(review1, HOLD)
+        handler.postModerationRequest(review1, ProductReviewStatus.HOLD)
         advanceTimeBy(100)
         runCurrent()
 
         // WHEN
-        handler.postModerationRequest(review2, HOLD)
+        handler.postModerationRequest(review2, ProductReviewStatus.HOLD)
         advanceTimeBy(100)
         runCurrent()
 
@@ -309,11 +267,6 @@
     fun `when adding a second moderation request, then queue it after the previous one`() = runTest {
         // GIVEN
         val statusUpdates = mutableListOf<List<ReviewModerationStatus>>()
-        handler = ReviewModerationHandler(
-            selectedSite = selectedSite,
-            productStore = productStore,
-            appCoroutineScope = backgroundScope,
-        )
 
         val review1 = review.copy(remoteId = 1L)
         val review2 = review.copy(remoteId = 2L)
@@ -327,8 +280,8 @@
         runCurrent()
 
         // WHEN
-        handler.postModerationRequest(review1, HOLD)
-        handler.postModerationRequest(review2, HOLD)
+        handler.postModerationRequest(review1, ProductReviewStatus.HOLD)
+        handler.postModerationRequest(review2, ProductReviewStatus.HOLD)
         advanceTimeBy(100)
         runCurrent()
 
@@ -342,20 +295,14 @@
     @Test
     fun `when queuing moderations, then make sure the pending status list is ordered`() = runTest {
         // GIVEN
-        handler = ReviewModerationHandler(
-            selectedSite = selectedSite,
-            productStore = productStore,
-            appCoroutineScope = backgroundScope,
-        )
-
         val review1 = review.copy(remoteId = 1L)
         val review2 = review.copy(remoteId = 2L)
 
         runCurrent()
 
         // WHEN
-        handler.postModerationRequest(review1, HOLD)
-        handler.postModerationRequest(review2, HOLD)
+        handler.postModerationRequest(review1, ProductReviewStatus.HOLD)
+        handler.postModerationRequest(review2, ProductReviewStatus.HOLD)
         advanceTimeBy(100)
         runCurrent()
 
@@ -369,7 +316,11 @@
     @Test
     fun `when moderating to different review statuses, then handle each correctly`() = runTest {
         // GIVEN
-        val statuses = listOf(APPROVED, SPAM, TRASH)
+        val statuses = listOf(
+            ProductReviewStatus.APPROVED,
+            ProductReviewStatus.SPAM,
+            ProductReviewStatus.TRASH
+        )
 
         for (status in statuses) {
             // GIVEN
@@ -382,12 +333,6 @@
                 )
             )
 
-            handler = ReviewModerationHandler(
-                selectedSite = selectedSite,
-                productStore = productStore,
-                appCoroutineScope = backgroundScope,
-            )
-
             runCurrent()
 
             // WHEN
@@ -408,11 +353,6 @@
     fun `when posting multiple requests for same review rapidly, then only process the latest`() = runTest {
         // GIVEN
         val statusUpdates = mutableListOf<List<ReviewModerationStatus>>()
-        handler = ReviewModerationHandler(
-            selectedSite = selectedSite,
-            productStore = productStore,
-            appCoroutineScope = backgroundScope,
-        )
 
         val collectJob = launch {
             handler.pendingModerationStatus.collect { statuses ->
@@ -423,9 +363,9 @@
         runCurrent()
 
         // WHEN
-        handler.postModerationRequest(review, HOLD)
-        handler.postModerationRequest(review, APPROVED)
-        handler.postModerationRequest(review, SPAM)
+        handler.postModerationRequest(review, ProductReviewStatus.HOLD)
+        handler.postModerationRequest(review, ProductReviewStatus.APPROVED)
+        handler.postModerationRequest(review, ProductReviewStatus.SPAM)
         runCurrent()
         advanceTimeBy(100)
         runCurrent()
@@ -438,18 +378,12 @@
         }
 
         val latestPending = pendingForSameReview.lastOrNull()
-        assertThat(latestPending?.newStatus).isEqualTo(SPAM)
+        assertThat(latestPending?.newStatus).isEqualTo(ProductReviewStatus.SPAM)
     }
 
     @Test
     fun `when undoing operation that doesn't exist, then handle gracefully`() = runTest {
         // GIVEN
-        handler = ReviewModerationHandler(
-            selectedSite = selectedSite,
-            productStore = productStore,
-            appCoroutineScope = backgroundScope,
-        )
-
         runCurrent()
         val nonExistentReview = review.copy(remoteId = 999L)
 
@@ -466,11 +400,6 @@
     fun `when handling concurrent operations on different reviews, then process all correctly`() = runTest {
         // GIVEN
         val statusUpdates = mutableListOf<List<ReviewModerationStatus>>()
-        handler = ReviewModerationHandler(
-            selectedSite = selectedSite,
-            productStore = productStore,
-            appCoroutineScope = backgroundScope,
-        )
 
         val review1 = review.copy(remoteId = 1L)
         val review2 = review.copy(remoteId = 2L)
@@ -485,9 +414,9 @@
         runCurrent()
 
         // WHEN
-        handler.postModerationRequest(review1, APPROVED)
-        handler.postModerationRequest(review2, SPAM)
-        handler.postModerationRequest(review3, TRASH)
+        handler.postModerationRequest(review1, ProductReviewStatus.APPROVED)
+        handler.postModerationRequest(review2, ProductReviewStatus.SPAM)
+        handler.postModerationRequest(review3, ProductReviewStatus.TRASH)
         runCurrent()
         advanceTimeBy(100)
         runCurrent()
```

</details>

FYI: There's only this last test that fails when you'll apply this patch, and for the life of me, I just couldn't fix it without altering the assert logic at the very end, maybe you can fix it... 🤞 

PS: Again, many thanks for this! 🙇 ❤️ 🚀

malinajirka committed:
Clean up reviews tests and make them more robust

malinajirka commented:
Noice, thank you for the review as well as the patch! That's definitely much nicer ❤️.

malinajirka committed:
Revert "Clean up reviews tests and make them more robust"

This reverts commit 69fc6115aea1e8cb4c6d468fc8b7df7c59feda59.

malinajirka committed:
Revert back to using backgroundScope

malinajirka commented:
> FYI: There's only this last test that fails when you'll apply this patch, and for the life of me, I just couldn't fix it without altering the assert logic at the very end, maybe you can fix it... 🤞

Ahh, I missed this part of your comment and I applied the patch without thinking about it more deeply.

1. Initializing the handler on the class level makes it shared across all the tests which is a bit risky. At first I thought moving the initializing to @Before method would work, but it won't because of `2`.
2. I don't have 100% clarity on this, so take this with a grain of salt. But I think the issue with the patch is that we create a TestScope which is unrelated to the test scope created by `runTest`. Moreover, I think the usage of `backgroundScope` is correct here since the production code uses Application Level scope and AFAIU that's exactly what `backgroundScope` is for.

I extracted creation of the SUT to a method, but I reverted back to calling it in each method to make sure we use scope of the `runTest`. 

Let me know what you think @ParaskP7

ParaskP7 commented:
👋 @malinajirka !

> Ahh, I missed this part of your comment and I applied the patch without thinking about it more deeply.

😊 yea, I should have made that more clear, apologies... 🫣

> 1. Initializing the handler on the class level makes it shared across all the tests which is a bit risky.

Actually, the interesting thing about this and unit tests, is that using this `val` like that, actually any test class field, would not be shared across all the tests, this is a common misconception which I too forget from time-to-time.

FYI: At the property level, this handler will be initialized when the test class instance is created.
- JUnit (and typical JVM test frameworks) create a new instance of a test class for each @Test method.
- Therefore, for each test, a fresh instance of a test class is created, and a property initializer runs for that instance.
- ...and thus:
    - `handler` is not shared between tests.
    - It’s initialized per test method, just like it would be if you set it up in a `@Before`-annotated function.
    - Each test gets a new handler.

PS: If one however declares a `handler` as `companion object`, `object`, or `@JvmStatic` then this would be shared across all test methods. But a regular property (even with private val) will be re-initialized per test method.

> At first I thought moving the initializing to https://github.com/before method would work, but it won't because of 2.

Yea, I tried that too, using `backgroundScope` on that `@Before` `runTest { ... }` level is not the same as using `backgroundScope` on a `@Test` `runTest { ... }`, I struggled with that myself... 😊

> 2. I don't have 100% clarity on this, so take this with a grain of salt. But I think the issue with the patch is that we create a TestScope which is unrelated to the test scope created by runTest.

Yea, I think you're right here and we should indeed use the test scope created by `runTest` to do what we want with these tests. 🤔

> Moreover, I think the usage of backgroundScope is correct here since the production code uses Application Level scope and AFAIU that's exactly what backgroundScope is for.

Yea, I haven't see usage of this `backgroundScope` much, so this through me off. I also didn't see it used throughout the codebase so it made me think if that's the correct approach. I really have limited knowledge about `backgroundScope` and where you really need to use that.

Thus, I proceeded with our usual `TestScope(coroutinesTestRule.testDispatcher)` approach and see how far I'll get with it, thinking that I am missing something on the last test that you could simply fix (or update accordingly). 😊

> I extracted creation of the SUT to a method, but I reverted back to calling it in each method to make sure we use scope of the runTest.

I do think we should indeed accept this solution and forward, apologies for all the back-and-forward! 👍  💯 😊

malinajirka commented:
> JUnit (and typical JVM test frameworks) create a new instance of a test class for each @test method.

TIL very interesting - I thought it calls Before on each test but didn't realize it creates a new instance of the test class 🤔.

> I also didn't see it used throughout the codebase so it made me think if that's the correct approach.

Might not be the correct approach, however, I think this case is a bit specific since we are injecting app level scope into the SUT instead of just dispatchers as we do in vast majority of the codebase.

> I do think we should indeed accept this solution and forward, apologies for all the back-and-forward! 👍 💯 😊

No reason to apologize, I learned something new so definitely worth it! 💟

ParaskP7 commented:
> TIL very interesting - I thought it calls Before on each test but didn't realize it creates a new instance of the test class 🤔.

Yea, I know, this is one of the TILs I get myself multiple times a couple of years, as I forget that and re-learn that again, not intuitive at all... 🤷 

> Might not be the correct approach, however, I think this case is a bit specific since we are injecting app level scope into the SUT instead of just dispatchers as we do in vast majority of the codebase.

💯 it's indeed a specific case! 👍

> No reason to apologize, I learned something new so definitely worth it! 💟

🙇 ❤️ 🚀
